server/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -125,16 +126,32 @@ public GroupShardsIterator<ShardIterator> searchShards(
                 nodeCounts
             );
             if (iterator != null) {
-                var shardsThatCanHandleSearches = iterator.getShardRoutings()
-                    .stream()
-                    .filter(shardRouting -> canSearchShard(shardRouting, clusterState))
-                    .toList();
+                final List<ShardRouting> shardsThatCanHandleSearches;
+                if (isStateless) {
+                    shardsThatCanHandleSearches = statelessShardsThatHandleSearches(clusterState, iterator);
+                } else {
+                    shardsThatCanHandleSearches = statefulShardsThatHandleSearches(iterator);
+                }
                 set.add(new PlainShardIterator(iterator.shardId(), shardsThatCanHandleSearches));
             }
         }
         return GroupShardsIterator.sortAndCreate(new ArrayList<>(set));
     }
 
+    private static List<ShardRouting> statefulShardsThatHandleSearches(ShardIterator iterator) {
+        final List<ShardRouting> shardsThatCanHandleSearches = new ArrayList<>(iterator.size());
+        for (ShardRouting shardRouting : iterator) {
+            if (shardRouting.isSearchable()) {
+                shardsThatCanHandleSearches.add(shardRouting);
+            }
+        }
+        return shardsThatCanHandleSearches;
+    }
+
+    private static List<ShardRouting> statelessShardsThatHandleSearches(ClusterState clusterState, ShardIterator iterator) {
+        return iterator.getShardRoutings().stream().filter(shardRouting -> canSearchShard(shardRouting, clusterState)).toList();
```

Comment:
nit Wouldn't it be simpler to have extended the original .filter() clause with an `if (isStateless)`, rather than creating new methods? Maybe you chose to use a simple for loop in stateful for performance reasons?

Reply:
Yea loops ftw here performance wise, streams are just so unpredictable and pretty much always slower for this kind of thing :)

Reply:
profile for the previous code, just to illustrate how annoying reasoning about streams is + all the virtual call overhead just doesn't go away no matter how hot the code gets seemingly.

<img width="1512" alt="image" src="https://github.com/elastic/elasticsearch/assets/6490959/4d285a16-5c46-4aa1-b4f9-78c19c2058f7">